### ISSUE TITLE
Simple Paged Search control needs to be used more sparingly

### ISSUE THREAD

**Issue by pbrezina** [pbrezina/sssd-test#1206]
Cloned from Pagure issue: https://pagure.io/SSSD/sssd/issue/1202
- Created at 2012-02-21 14:04:30 by [sgallagh](https://pagure.io/user/sgallagh)
- Closed as Fixed
- Assigned to [sgallagh](https://pagure.io/user/sgallagh)
---

We added support for the Simple Paged Search in ticket #658 and SSSD 1.5.8. When we did so, we chose to default to always requesting paged searches for all requests. However, it turns out that RFC 2696 does not mandate that the LDAP server supports more than one paging search at a time on the same connection. Some LDAP servers (notably 389 DS 1.2.9 and older) allow only a single paged search to be in progress on a request at one time. This results in receiving an error from LDAP that the server refuses to perform the operation. Once ticket #1201 is fixed, this will result in offline operation (right now it's misbehaving as if it received zero results). We need to avoid both of these events.

RFC 2696 does specify that if the result size being returned is smaller than the requested page size that the paging control **should** be ignored. This would mitigate the problem, but it appears that 389 DS also chooses to not ignore the control in this case.

In general, we need to find ways to minimize our usage of the paging control where available. I propose the following changes:

1. Add a boolean option to {{{ldap_generic_search_send()}}} to specify whether to use the paging control (when available).
1. Do not request the paging control when we expect a single-valued return from the request.
1. Always request the paging control for enumeration requests.
1. Serialize paging requests by maintaining a queue of lookups requiring paging and allow only one to operate at a time.{{{[1]}}}
1. If it can be reasonably determined that replies will never exceed 500 entries, also omit the paging control{{{[2]}}}

{{{[1]}}}: This will likely introduce apparent performance issues when performing initgroups() requests, since they will no longer be able to run in parallel. We must always assume that the user may belong to many groups.

{{{[2]}}}: I'm not 100% sure about this one. It's possible for a server to manually set its page size smaller than 500, though I doubt very many would do so. On the other hand, I can't come up with any ideas offhand for lookups of greater than a single entry that could not therefore result in a potentially-infinite set of results. So this may be a moot point.

### Comments

---

**Comment from [sgallagh](https://pagure.io/user/sgallagh) at [2012-02-22 02:25:54](https://pagure.io/SSSD/sssd/issue/1202#226682)**

Fields changed

status: new => assigned
version: 1.8.0 beta 2 => 1.8.0 beta 3

---

**Comment from [sgallagh](https://pagure.io/user/sgallagh) at [2012-02-22 02:26:03](https://pagure.io/SSSD/sssd/issue/1202#226683)**

Fields changed

milestone: NEEDS_TRIAGE => SSSD 1.8.0 (LTM)

---

**Comment from [sgallagh](https://pagure.io/user/sgallagh) at [2012-02-22 17:40:00](https://pagure.io/SSSD/sssd/issue/1202#226684)**

Ticket has been cloned to Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=796307

rhbz:  => [https://bugzilla.redhat.com/show_bug.cgi?id=796307 796307]

---

**Comment from [sgallagh](https://pagure.io/user/sgallagh) at [2012-02-22 19:50:02](https://pagure.io/SSSD/sssd/issue/1202#226685)**

Fields changed

blocking:  => 1207

---

**Comment from [sgallagh](https://pagure.io/user/sgallagh) at [2012-02-22 19:51:15](https://pagure.io/SSSD/sssd/issue/1202#226686)**

I have split this into two tickets. The mitigation pieces (1 and 2 above) will be resolved by this ticket. The more complete portion will be resolved in Ticket #1207

patch: 0 => 1

---

**Comment from [sgallagh](https://pagure.io/user/sgallagh) at [2012-02-24 21:06:05](https://pagure.io/SSSD/sssd/issue/1202#226687)**

Fixed by:
    - 2f3ee3f49019f5b60adbe073070f31e6e2d7c7ab (master)
    - 3828873b48096e6482329bab6da175de3f615ab8 (sssd-1-8)

resolution:  => fixed
status: assigned => closed

---

**Comment from [sgallagh](https://pagure.io/user/sgallagh) at [2012-03-14 12:55:13](https://pagure.io/SSSD/sssd/issue/1202#226688)**

Linked to Bugzilla bug: https://bugzilla.redhat.com/show_bug.cgi?id=782221 (*Red Hat Enterprise Linux 5*)

rhbz: [https://bugzilla.redhat.com/show_bug.cgi?id=796307 796307] => [https://bugzilla.redhat.com/show_bug.cgi?id=796307 796307], [https://bugzilla.redhat.com/show_bug.cgi?id=782221 782221]

---

**Comment from [sgallagh](https://pagure.io/user/sgallagh) at [2017-02-24 14:34:04](https://pagure.io/SSSD/sssd/issue/1202#226689)**

**Metadata Update from @sgallagh**:
- Issue assigned to sgallagh
- Issue marked as blocked by: #1207
- Issue set to the milestone: SSSD 1.8.0 (LTM)
